File: src/cli/commands/add-lifetour-offer.ts
import type { CommandHandler, CliContext } from '../shared/types';
import { registerCommand } from './registry';

/**
 * Specialized fast helper for Lifetour (五福) product/search pages.
 * Optimized for quick manual entry during Shaping Stage research.
 */
const addLifetourOfferCommand: CommandHandler = {
  names: ['add-lifetour-offer'],
  description: 'Fast path for Lifetour (五福) FIT/group offers. Writes directly to Turso (no JSON file).',
  usage: [
    'add-lifetour-offer --url <lifetour url> --price <twd> --hotel "<name>"',
    '                 [--depart YYYY-MM-DD] [--return YYYY-MM-DD] [--seats N] [--note "..."] [--run <run_id>]',
    '',
    'The URL can be either a search list page or a specific product page.',
    'Example:',
    '  add-lifetour-offer --url "https://tour.lifetour.com.tw/searchlist/tpe/0001-0005" \\',
    '    --price 17200 --hotel "水之都那霸飯店" --seats 2 --depart 2026-06-21 --return 2026-06-24',
  ].join('\n  '),
  requiresState: false,
  async execute(ctx: CliContext): Promise<void> {
    const { args } = ctx;

    const url = args.optionValue('--url');
    const priceStr = args.optionValue('--price');
    const hotel = args.optionValue('--hotel');

    if (!url || !priceStr || !hotel) {
      console.error('Error: --url, --price and --hotel are required');
      process.exit(1);
    }

    const price = parseInt(priceStr, 10);
    const seats = args.optionValue('--seats') ? parseInt(args.optionValue('--seats')!, 10) : null;
    const note = args.optionValue('--note') || '';
    const runId = args.optionValue('--run') || 'shaping-20260525-093508';

    // Try to detect Okinawa from the common Lifetour code
    let region = 'okinawa';
    if (url.includes('0001-0005')) {
      region = 'okinawa';
    } else if (url.includes('0001-0001')) {
      region = 'tokyo'; // or kanto
    } else if (url.includes('0001-0003')) {
      region = 'kansai';
    }

    const nightsStr = args.optionValue('--nights');
    let nights = nightsStr ? parseInt(nightsStr, 10) : 3;

    let depart = args.optionValue('--depart');
    let ret = args.optionValue('--return');

    // If user didn't provide dates, we still require them for now
    if (!depart || !ret) {
      console.error('Error: Please provide --depart and --return (YYYY-MM-DD)');
      console.error('Lifetour pages are usually search results, so dates are not embedded in the URL.');
      process.exit(1);
    }

    if (!nightsStr) {
      // Try to infer nights from return - depart if possible
      const d1 = new Date(depart);
      const d2 = new Date(ret);
      const diffDays = Math.round((d2.getTime() - d1.getTime()) / (1000 * 3600 * 24));
      nights = Math.max(1, diffDays - 1);
    }

    const title = `【機加酒．沖繩自由行${nights + 1}日】${hotel}、五福旅遊`;

    const now = new Date().toISOString();
    const offerId = `lifetour-okinawa-${depart.replace(/-/g, '')}-${nights}n-${Date.now().toString(36)}`;

    // De-JSON'd: typed columns for read fields, flat notes for the rest (no JSON).
    const notes: Array<{ key: string; value: string }> = [
      { key: 'source', value: 'lifetour_manual' },
      { key: 'url', value: url },
      { key: 'observed_at', value: now },
    ];

    const svc = require('../../services/tour-group-service');

    const row = {
      run_id: runId,
      offer_id: offerId,
      source_id: 'lifetour',
      dest_region: region,
      depart_date: depart,
      return_date: ret,
      nights,
      price_per_person_twd: price,
      title,
      url,
      scraped_at: now,
      hotel_name: hotel,
      hotel_star_rating: null,
      meals_included_count: 0,
      departure_status: 'available',
      seats_available: seats,
      min_group_size: 2,
      group_size_cap: null,
      raw_note: note || null,
      notes,
      product_kind: 'fit',
    };

    await svc.insertTourGroupOffers([row]);

    console.log(`✅ Lifetour offer saved directly to Turso`);
    console.log(`   ${depart} → ${ret} (${nights}n) | ${price} TWD/pax | ${hotel}`);
    console.log(`   run: ${runId}`);
    console.log(`   url: ${url}`);
  },
};

registerCommand(addLifetourOfferCommand);

File: rust/crates/travel-cli/src/add_lifetour_offer.rs
// `travel add-lifetour-offer --url <lifetour url> --price <TWD> --hotel "<name>" [--depart YYYY-MM-DD] [--return YYYY-MM-DD] [--seats <N>] [--note "<text>"] [--run <run_id>]`
//
// Port of src/cli/commands/add-lifetour-offer.ts.

use crate::cascade::common::now_rfc3339;
use crate::tour_group_offers::{insert_tour_group_offers, base36_timestamp, Note, TourGroupOfferRow};

pub async fn run(rest: &[String]) -> Result<(), String> {
    // First check for --help/-h
    if rest.iter().any(|a| a == "--help" || a == "-h") {
        println!("Usage:\n  travel add-lifetour-offer --url <lifetour url> --price <TWD> --hotel \"<name>\" [--depart YYYY-MM-DD] [--return YYYY-MM-DD] [--seats N] [--note \"<text>\"] [--run <run_id>]");
        return Ok(());
    }

    // Parse arguments
    let mut url = None;
    let mut price_str = None;
    let mut hotel = None;
    let mut depart = None;
    let mut ret = None;
    let mut nights_str = None;
    let mut seats_str = None;
    let mut note = None;
    let mut run_id = None;

    let mut i = 0;
    while i < rest.len() {
        match rest[i].as_str() {
            "--url" => {
                i += 1;
                url = rest.get(i).cloned();
            }
            "--price" => {
                i += 1;
                price_str = rest.get(i).cloned();
            }
            "--hotel" => {
                i += 1;
                hotel = rest.get(i).cloned();
            }
            "--depart" => {
                i += 1;
                depart = rest.get(i).cloned();
            }
            "--return" => {
                i += 1;
                ret = rest.get(i).cloned();
            }
            "--nights" => {
                i += 1;
                nights_str = rest.get(i).cloned();
            }
            "--seats" => {
                i += 1;
                seats_str = rest.get(i).cloned();
            }
            "--note" => {
                i += 1;
                note = rest.get(i).cloned();
            }
            "--run" => {
                i += 1;
                run_id = rest.get(i).cloned();
            }
            other if other.starts_with("--") => {
                return Err(format!("unknown argument: {other}"));
            }
            _ => {}
        }
        i += 1;
    }

    // Validate required fields
    let Some(url) = url else {
        eprintln!("Error: --url, --price and --hotel are required");
        std::process::exit(1);
    };
    let Some(price_str) = price_str else {
        eprintln!("Error: --url, --price and --hotel are required");
        std::process::exit(1);
    };
    let Some(hotel) = hotel else {
        eprintln!("Error: --url, --price and --hotel are required");
        std::process::exit(1);
    };

    let price = price_str.parse::<i64>().map_err(|e| format!("invalid price: {e}"))?;
    let seats = seats_str.and_then(|s| s.parse::<i64>().ok());
    let run_id = run_id.unwrap_or_else(|| "shaping-20260525-093508".to_string());

    // Determine region from URL
    let mut region = "okinawa".to_string();
    if url.contains("0001-0005") {
        region = "okinawa".to_string();
    } else if url.contains("0001-0001") {
        region = "tokyo".to_string();
    } else if url.contains("0001-0003") {
        region = "kansai".to_string();
    }

    // Determine nights
    let mut nights = nights_str.as_ref().and_then(|s| s.parse::<i64>().ok()).unwrap_or(3);

    // Determine dates
    let Some(depart) = depart else {
        eprintln!("Error: Please provide --depart and --return (YYYY-MM-DD)");
        eprintln!("Lifetour pages are usually search results, so dates are not embedded in the URL.");
        std::process::exit(1);
    };
    let Some(ret) = ret else {
        eprintln!("Error: Please provide --depart and --return (YYYY-MM-DD)");
        eprintln!("Lifetour pages are usually search results, so dates are not embedded in the URL.");
        std::process::exit(1);
    };

    // Infer nights from dates if not provided
    if nights_str.is_none()
        && let (Ok(depart_date), Ok(ret_date)) = (
            chrono::NaiveDate::parse_from_str(&depart, "%Y-%m-%d"),
            chrono::NaiveDate::parse_from_str(&ret, "%Y-%m-%d"),
        )
    {
        let diff = ret_date.signed_duration_since(depart_date);
        let diff_days = diff.num_days();
        nights = std::cmp::max(1, diff_days - 1);
    }

    let now = now_rfc3339();

    let depart_no_dash = depart.replace('-', "");
    let offer_id = format!("lifetour-okinawa-{}-{}n-{}", depart_no_dash, nights, base36_timestamp());

    let title = format!("【機加酒．沖繩自由行{}日】{}、五福旅遊", nights + 1, hotel);

    let notes = vec![
        Note { key: "source".to_string(), value: "lifetour_manual".to_string() },
        Note { key: "url".to_string(), value: url.clone() },
        Note { key: "observed_at".to_string(), value: now.clone() },
    ];

    let row = TourGroupOfferRow {
        run_id: run_id.clone(),
        offer_id: offer_id.clone(),
        source_id: "lifetour".to_string(),
        dest_region: region,
        depart_date: depart.clone(),
        return_date: ret.clone(),
        nights,
        price_per_person_twd: price,
        title,
        url: url.clone(),
        scraped_at: now,
        hotel_name: Some(hotel.clone()),
        hotel_star_rating: None,
        meals_included_count: Some(0),
        departure_status: Some("available".to_string()),
        seats_available: seats,
        min_group_size: Some(2),
        group_size_cap: None,
        raw_confidence: None,
        raw_note: note,
        raw_flight: None,
        raw_flight_outbound: None,
        raw_flight_return: None,
        notes,
        product_kind: Some("fit".to_string()),
    };

    // Connect to DB and insert
    let conn = crate::db::connect_write().await?;
    insert_tour_group_offers(&conn, &[row]).await?;

    println!("✅ Lifetour offer saved directly to Turso");
    println!("   {depart} → {ret} ({nights}n) | {price} TWD/pax | {hotel}");
    println!("   run: {run_id}");
    println!("   url: {url}");

    Ok(())
}
